### PR TITLE
LPS-153034 Update Layout API to support addition of Categorization Framework

### PR DESCRIPTION
**Pull request by seltong**

### CODE REVIEW

**Comment by seltong** [comment]
ci:test:sf

**Comment by seltong** [comment]
Just started reviewing :)

File: modules/apps/object/object-api/src/main/java/com/liferay/object/exception/ObjectLayoutBoxCategorizationTypeException.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package com.liferay.object.exception;
+
+import com.liferay.portal.kernel.exception.PortalException;
+
+/**
+ * @author Selton Guedes
+ */
+public class ObjectLayoutBoxCategorizationTypeException
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/exception/ObjectLayoutBoxCategorizationTypeException.java:22]: Hi @seltong, this exception must be autogenerated, look for exception in service.xml and follow the pattern.

Thank you!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectLayoutLocalServiceImpl.java [hunk 2]
```diff
@@ -235,7 +242,7 @@ public ObjectLayout updateObjectLayout(
 	private ObjectLayoutBox _addObjectLayoutBox(
 			User user, long objectDefinitionId, long objectLayoutTabId,
 			boolean collapsable, Map<Locale, String> nameMap, int priority,
-			List<ObjectLayoutRow> objectLayoutRows)
+			List<ObjectLayoutRow> objectLayoutRows, String objectLayoutBoxType)
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectLayoutLocalServiceImpl.java:245]: Hi @seltong, sort this variable, please.

Thanks!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectLayoutLocalServiceImpl.java [hunk 3]
```diff
@@ -567,6 +576,44 @@ private void _validate(
 		}
 	}
 
+	private void _validateObjectLayoutBoxCategorizationType(
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectLayoutLocalServiceImpl.java:579]: Hi @seltong, I believe that you can use the _validate method to check the restricions of categorization objectLayoutBox.

File: modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectLayoutResourceImpl.java [hunk 4]
```diff
@@ -170,6 +171,13 @@ private com.liferay.object.model.ObjectLayoutBox _toObjectLayoutBox(
 
 		String objectLayoutBoxType = objectLayoutBox.getTypeAsString();
 
+		if (!GetterUtil.getBoolean(PropsUtil.get("feature.flag.LPS-149014")) &&
+			(objectLayoutBoxType != null)) {
+
+			throw new UnsupportedOperationException(
+				"Categorization layout box is not yet supported");
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectLayoutResourceImpl.java:178]: Hi @seltong, Brian prefers to throw this exception without a message, so just replace to `throw new UnsupportedOperationException();`

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

**Comment by seltong** [comment]
ci:test:sf

File: modules/apps/object/object-api/bnd.bnd [hunk 5]
```diff
@@ -1,6 +1,6 @@
 Bundle-Name: Liferay Object API
 Bundle-SymbolicName: com.liferay.object.api
-Bundle-Version: 20.1.1
+Bundle-Version: 20.2.0
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-api/bnd.bnd:3]: Hi @seltong, please, move move this change to semver commit

Thank you!

File: modules/apps/object/object-api/src/main/resources/com/liferay/object/constants/packageinfo [hunk 6]
```diff
@@ -1 +1 @@
-version 2.7.0
\ No newline at end of file
+version 2.8.0
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-api/src/main/resources/com/liferay/object/constants/packageinfo:1]: This one and the other below as well.

Thanks!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectLayoutLocalServiceImpl.java [hunk 7]
```diff
@@ -233,8 +231,9 @@ public ObjectLayout updateObjectLayout(
 	}
 
 	private ObjectLayoutBox _addObjectLayoutBox(
-			User user, long objectDefinitionId, long objectLayoutTabId,
-			boolean collapsable, Map<Locale, String> nameMap, int priority,
+			User user, long objectDefinitionId, String objectLayoutBoxType,
```
Comment (MarinhoFeliphe) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectLayoutLocalServiceImpl.java:234]: Hi @seltong, just pass objectLayoutBox as parameter and get its propeties.

for example:


			long objectDefinitionId, ObjectLayoutBox objectLayoutBox,
			long objectLayoutTabId, User user)
		throws PortalException {

		ObjectLayoutBox newObjectLayoutBox = _objectLayoutBoxPersistence.create(
			counterLocalService.increment());

		newObjectLayoutBox.setCompanyId(user.getCompanyId());
		newObjectLayoutBox.setUserId(user.getUserId());
		newObjectLayoutBox.setUserName(user.getFullName());
		newObjectLayoutBox.setObjectLayoutTabId(objectLayoutTabId);
		newObjectLayoutBox.setCollapsable(objectLayoutBox.isCollapsable());
		newObjectLayoutBox.setNameMap(objectLayoutBox.getNameMap());
		newObjectLayoutBox.setPriority(objectLayoutBox.getPriority());
            newObjectLayoutBox.setType(objectLayoutBox.getType());

		newObjectLayoutBox =
			_objectLayoutBoxPersistence.update(newObjectLayoutBox);

		newObjectLayoutBox.setObjectLayoutRows(
			_addObjectLayoutRows(
				user, objectDefinitionId,
				newObjectLayoutBox.getObjectLayoutBoxId(),
				objectLayoutBox.getObjectLayoutRows()));

		return newObjectLayoutBox;
	}

Thanks!

File: modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectLayoutResourceImpl.java (hunk 4 above)
Comment (MarinhoFeliphe) [inline modules/apps/object/object-admin-rest-impl/src/main/java/com/liferay/object/admin/rest/internal/resource/v1_0/ObjectLayoutResourceImpl.java:178]: Hi @seltong, take a look at this.

**Review by MarinhoFeliphe — CHANGES_REQUESTED** [review]

**Comment by seltong** [comment]
Just started reviewing :)

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:forward

**Comment by MariMalta** [comment]
Just started reviewing :)

**Comment by MariMalta** [comment]
Just started reviewing :)

**Comment by seltong** [comment]
ci:stop

**Comment by MariMalta** [comment]
Just started reviewing :)

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:forward

**Comment by seltong** [comment]
ci:test:relevant

**Comment by seltong** [comment]
ci:forward

**Comment by naraandrade** [comment]
Just started reviewing :)

**Pull request closed without merging**